custom-agent-demo/src/components/AgentDemo.js

```diff
@@ -0,0 +1,145 @@
+import React, { Component } from 'react'
+import { Device, PlatformIcon, SmartConnectButton } from 'cobrowse-agent-ui'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faTrash, faDoorOpen, faPen, faPencilAlt, faHandPointer } from '@fortawesome/free-solid-svg-icons'
+import CobrowseAPI from 'cobrowse-agent-sdk'
+import Button from './Button'
+import './AgentDemo.css'
+
+const cobrowse = new CobrowseAPI()
+let context = null
+
+export default class AgentDemo extends Component {
+  constructor () {
+    super()
+    const query = new URLSearchParams(window.location.search)
+    this.state = {
+      demo_id: query.get('demo_id'),
+      device_id: null,
+      devices: [],
+      token: query.get('token')
+    }
+
+    cobrowse.token = this.state.token
+  }
+
+  async getDevices () {
+    const devices = await cobrowse.devices.list()
+
+    if (Array.isArray(devices)) {
+      devices.forEach(device => {
+        device.subscribe()
+        device.on('updated', () => {
+          this.setState({ state: this.state })
+        })
+      })
+    }
+
+    this.setState({ devices })
+  }
+
+  renderDevices () {
+    const { devices } = this.state
+
+    if (devices.length > 0) {
+      return devices.map(device => {
+        return (
+          <div key={device.id} style={{ border: `2px solid ${device.online ? '#3cc56e' : '#eee'}`, borderRadius: 5, marginTop: 7, marginLeft: 7, marginRight: 7 }}>
+            <Device style={{border: '0px none', marginTop: 7}} device={device}>
+            <SmartConnectButton
+              key={device.id}
+              label={
+                <div style={{display:'flex', alignItems:'center'}}>
+                  <PlatformIcon style={{width: 20, height: 20, marginLeft:0}} platform={device.device.platform} />
+                  <span style={{marginLeft:7}}>Join</span>
+                </div>
+              }
+              onClick={() => this.connect(device.id)}
+              device={device}
+            />
+            </Device>
+            { Object.keys(device.custom_data).map(key => (
+              <div style={{ display: 'inline-block', margin: 4, padding: '3px 7px', fontSize: 12, borderRadius: 10, background:'#f3f3f3' }} key={key}>
+                <b>{key}</b><span> = </span><code>{device.custom_data[key]}</code>
+              </div>
+            ))}
+          </div>
+        )
+      })
+    } else {
+      this.getDevices()
```

Comment:
render methods should not have side effects

Reply:
@headlessme I'm not sure that this is a side effect. getDevices is an async pattern; this allows renderDevices to do a non-blocking initial render and change state if there are devices to render, otherwise it remains null.

Can do in componentDidMount as well, but that could be too late; if the promise resolves before the state is applied, devices will be rendered before componentDidMount, which is desirable in terms of the rendering lifecycle because it results in fewer renders.

Reply:
In this case, anything that isn't rendering JSX is a side effect. React render functions should be pure functions of state (this.state or this.props) wherever possible. The side effect here is it starts an async process of fetching which is nothing to do with converting state to virtual DOM elements. 

If the devices aren't loaded into `this.state yet`, the render method should either return null, or return some loading indicator. See https://reactjs.org/docs/hooks-effect.html for the most hipster way to do side effects.

Also, `render()` might get called many times before `getDevices()` finished fetching its data, in which case it will trigger a whole series of `getDevices()` calls.

Reply:
I didn't see it as breaking from the responsibility of converting from JSX to DOM.

Your point about multiple renders makes that clearly a strong division of concerns, thanks. Will move to componentDidMount. Will also add to componentDidUpdate for when sessions end.